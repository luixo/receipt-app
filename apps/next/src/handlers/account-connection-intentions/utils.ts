import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
