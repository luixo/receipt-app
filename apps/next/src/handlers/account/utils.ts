import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
