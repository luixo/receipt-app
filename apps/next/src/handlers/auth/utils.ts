import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
