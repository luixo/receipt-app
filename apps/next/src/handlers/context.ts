import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
