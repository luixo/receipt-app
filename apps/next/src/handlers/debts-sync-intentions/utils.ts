import * as trpc from "@trpc/server";
import { z } from "zod";

import { Database } from "next-app/db";
import { AccountsId, DebtsId } from "next-app/db/models";
import { getValidParticipants } from "next-app/handlers/receipt-items/utils";

export const getDebtIntention = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
) =>
	database
		.selectFrom("debts")
		.where("debts.id", "=", debtId)
		.leftJoin("debts as debtsMine", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsMine.id")
				.on("debtsMine.ownerAccountId", "=", accountId),
		)
		.leftJoin("debtsSyncIntentions", (qb) =>
			qb.onRef("debts.id", "=", "debtsSyncIntentions.debtId"),
		)
		.leftJoin("debts as debtsTheir", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsTheir.id")
				.on("debtsTheir.ownerAccountId", "<>", accountId),
		)
		.select([
			"debtsSyncIntentions.ownerAccountId as intentionAccountId",
			"debtsMine.lockedTimestamp as mineLockedTimestamp",
			"debtsMine.ownerAccountId as mineOwnerAccountId",
			"debtsTheir.lockedTimestamp as theirLockedTimestamp",
			"debtsTheir.ownerAccountId as theirOwnerAccountId",
		])
		.executeTakeFirst();

const syncStatusSchema = z.discriminatedUnion("type", [
	// The foreign user doesn't have a debt
	z.strictObject({ type: z.literal("nosync") }),
	// The foreign user debt's data is different
	z.strictObject({
		type: z.literal("unsync"),
		intention: z
			.strictObject({
				direction: z.union([z.literal("self"), z.literal("remote")]),
			})
			.optional(),
	}),
	// The foreign user debt's data is in sync
	z.strictObject({ type: z.literal("sync") }),
]);

export type SyncStatus = z.infer<typeof syncStatusSchema>;

export const getSyncStatus = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
): Promise<SyncStatus> => {
	const debtIntention = await getDebtIntention(database, debtId, accountId);
	if (!debtIntention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Debt ${debtId} does not exist.`,
		});
	}
	const {
		mineOwnerAccountId,
		mineLockedTimestamp,
		theirOwnerAccountId,
		theirLockedTimestamp,
		intentionAccountId,
	} = debtIntention;
	if (!mineOwnerAccountId) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: `Debt ${debtId} does not exist for user ${accountId}.`,
		});
	}
	if (
		theirLockedTimestamp &&
		mineLockedTimestamp &&
		theirLockedTimestamp.valueOf() === mineLockedTimestamp.valueOf()
	) {
		return { type: "sync" };
	}
	if (!intentionAccountId) {
		if (theirOwnerAccountId) {
			return { type: "unsync" };
		}
		return { type: "nosync" };
	}
	if (intentionAccountId === accountId) {
		return {
			type: "unsync",
			intention: { direction: "self" },
		};
	}
	return {
		type: "unsync",
		intention: { direction: "remote" },
	};
};

type Participant = Awaited<ReturnType<typeof getValidParticipants>>[number];

export const upsertDebtIntentionFromReceipt = async (
	database: Database,
	participants: Participant[],
	ownerAccountId: AccountsId,
	created: Date,
) =>
	database
		.insertInto("debtsSyncIntentions")
		.values(
			participants.map((participant) => ({
				debtId: participant.debtId,
				ownerAccountId,
				lockedTimestamp: created,
			})),
		)
		.onConflict((oc) =>
			oc.column("debtId").doUpdateSet({
				ownerAccountId: (eb) => eb.ref("excluded.ownerAccountId"),
				lockedTimestamp: (eb) => eb.ref("excluded.lockedTimestamp"),
			}),
		)
		.execute();
