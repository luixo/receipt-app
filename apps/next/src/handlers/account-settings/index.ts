import { t } from "next-app/handlers/trpc";

import { procedure as get } from "./get";
import { procedure as update } from "./update";

export const router = t.router({
	get,
	update,
});
