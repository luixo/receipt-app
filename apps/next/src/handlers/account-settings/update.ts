import { z } from "zod";

import type { SimpleUpdateObject } from "next-app/db/types";
import { authProcedure } from "next-app/handlers/trpc";

import { DEFAULT_ACCOUNT_SETTINGS } from "./get";

type SettingsUpdateObject = SimpleUpdateObject<"accountSettings">;

export const procedure = authProcedure
	.input(
		z.discriminatedUnion("type", [
			z.strictObject({
				type: z.literal("autoAcceptDebts"),
				value: z.boolean(),
			}),
		]),
	)
	.mutation(async ({ ctx, input }) => {
		const { database } = ctx;
		const updateObject: SettingsUpdateObject = {};
		switch (input.type) {
			case "autoAcceptDebts":
				updateObject.autoAcceptDebts = input.value;
		}
		const existingSettings = await database
			.selectFrom("accountSettings")
			.where("accountSettings.accountId", "=", ctx.auth.accountId)
			.executeTakeFirst();
		if (!existingSettings) {
			await database
				.insertInto("accountSettings")
				.values({
					accountId: ctx.auth.accountId,
					...DEFAULT_ACCOUNT_SETTINGS,
					...updateObject,
				})
				.executeTakeFirst();
		} else {
			await database
				.updateTable("accountSettings")
				.set(updateObject)
				.where("accountSettings.accountId", "=", ctx.auth.accountId)
				.executeTakeFirst();
		}
	});
