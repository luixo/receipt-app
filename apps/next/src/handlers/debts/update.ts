import * as trpc from "@trpc/server";
import { MutationObject } from "kysely";
import { z } from "zod";

import { debtNoteSchema } from "app/utils/validation";
import { ReceiptsDatabase, getDatabase } from "next-app/db";
import { AuthorizedContext } from "next-app/handlers/context";
import { getDebt } from "next-app/handlers/debts/utils";
import {
	debtAmountSchema,
	currencySchema,
	debtIdSchema,
} from "next-app/handlers/validation";

export const router = trpc.router<AuthorizedContext>().mutation("update", {
	input: z.strictObject({
		id: debtIdSchema,
		update: z.discriminatedUnion("type", [
			z.strictObject({
				type: z.literal("amount"),
				amount: debtAmountSchema,
			}),
			z.strictObject({
				type: z.literal("timestamp"),
				timestamp: z.date(),
			}),
			z.strictObject({
				type: z.literal("note"),
				note: debtNoteSchema,
			}),
			z.strictObject({
				type: z.literal("currency"),
				currency: currencySchema,
			}),
		]),
	}),
	resolve: async ({ input, ctx }) => {
		const database = getDatabase(ctx);
		const debt = await getDebt(database, input.id, ctx.auth.accountId, []);
		if (!debt) {
			throw new trpc.TRPCError({
				code: "NOT_FOUND",
				message: `Debt ${input.id} does not exist on account ${ctx.auth.accountId}.`,
			});
		}
		let setObject: MutationObject<ReceiptsDatabase, "debts", "debts"> = {};
		switch (input.update.type) {
			case "amount":
				setObject = { amount: input.update.amount.toString() };
				break;
			case "timestamp":
				setObject = { timestamp: input.update.timestamp };
				break;
			case "note":
				setObject = { note: input.update.note };
				break;
			case "currency":
				setObject = { currency: input.update.currency };
				break;
		}
		await getDatabase(ctx)
			.updateTable("debts")
			.set(setObject)
			.where("id", "=", input.id)
			.executeTakeFirst();
	},
});
