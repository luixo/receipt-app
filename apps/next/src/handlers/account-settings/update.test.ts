import { describe } from "vitest";

import { t } from "next-app/handlers/trpc";
import { createAuthContext } from "next-tests/utils/context";
import {
	insertAccount,
	insertAccountSettings,
	insertAccountWithSession,
} from "next-tests/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectUnauthorizedError,
} from "next-tests/utils/expect";
import { test } from "next-tests/utils/test";

import { procedure } from "./update";

const router = t.router({ procedure });

describe("accountSettings.update", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			router
				.createCaller(context)
				.procedure({ type: "autoAcceptDebts", value: true }),
		);
	});

	describe("functionality", () => {
		describe("autoAcceptDebts", () => {
			test("settings not found - updated with default-like", async ({
				ctx,
			}) => {
				// Verifying other accounts are not affected
				await insertAccount(ctx);
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({ type: "autoAcceptDebts", value: false }),
				);
			});

			test("settings not found - updated with non-default value", async ({
				ctx,
			}) => {
				// Verifying other accounts are not affected
				await insertAccount(ctx);
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({ type: "autoAcceptDebts", value: true }),
				);
			});

			test("settings found", async ({ ctx }) => {
				// Verifying other accounts are not affected
				await insertAccount(ctx);
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				await insertAccountSettings(ctx, accountId, { autoAcceptDebts: true });
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({ type: "autoAcceptDebts", value: false }),
				);
			});
		});
	});
});
