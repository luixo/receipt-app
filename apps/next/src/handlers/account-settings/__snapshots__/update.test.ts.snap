// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`accountSettings.update > functionality > autoAcceptDebts > settings found 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public."accountSettings" VALUES ('535c16d9-e7f8-46b9-8ec1-7249f8db6e66', true);
+ INSERT INTO public."accountSettings" VALUES ('535c16d9-e7f8-46b9-8ec1-7249f8db6e66', false);"
`;

exports[`accountSettings.update > functionality > autoAcceptDebts > settings not found - updated with default-like 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
+ INSERT INTO public."accountSettings" VALUES ('cd58f9d1-e657-491e-b40b-d55c1cdcd984', false);"
`;

exports[`accountSettings.update > functionality > autoAcceptDebts > settings not found - updated with non-default value 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
+ INSERT INTO public."accountSettings" VALUES ('47376ecf-e921-41c2-8dfd-f733c8d46ed4', true);"
`;
